doc/aggr.1
.\" Process this file with
.\" groff -man -Tascii xml2flat.1
.\" or view with man -l xml2flat.1
.TH aggr 1 "NOVEMBER 2010"

.SH NAME
aggr \- aggregate CSV-data and print on the standard output

.SH SYNOPSIS
aggr [OPTIONS] COLUMNS...

.SH DESCRIPTION
aggr aggregates CSV-data (comma seperated values) from a file or standard input thru well-known aggregate functions and outputs the result to standard output. It tries to simulate some of the functionality that is provided by SQL's "GROUP BY"-statement.

.SH OPTIONS

.IP "-f INPUTFILE"
read data from INPUTFILE instead of standard input

.IP "-d DELIMITER"
use DELIMITER instead of Tab as field delimiter (for input and output)

.SH COLUMNS

The columns that will be printed in output are defined by the COLUMNS arguments. There are two kinds of columns:

.IP "Grouping columns"
define which columns are used to group the input data. They are simply specified by their column number. These columns are the attributes that one would mention in the GROUP BY statement in SQL.

.IP "Aggregated columns"
are columns whose values are calculated by an aggregate function. The syntax is AGGREGATEFUNCTION:COLUMNNUMBER. See below for supported aggregate functions.

.PP
The columns will appear in the output in the same order as in which they were specified in the arguments. It is possible to output the same column several times (with different aggregate functions). The column numbering scheme is 0-based: the first column has column number 0, the second 1 and so on.


.SH AGGREGATE FUNCTIONS

These aggregate functions are supported: count, sum, avg, min, max, first, last, median.


.SH EXAMPLES

Let's say there is a file name_city_age.csv with the following content (list of persons with name, city and age):

Hans;Regensburg;42
.br
Heidi;Straubing;40
.br
Peter;Straubing;20
.br
Else;Regensburg;10
.br
Hans;Augsburg;63
.br
Hans;Ingolstadt;71

Then

aggr -f name_city_age.csv -d ";" 1 avg:2

will return:

Regensburg;26.0
.br
Straubing;30.0
.br
Augsburg;63.0
.br
Ingolstadt;71.0

The grouping-criteria is the second column (column number 1) which is the column with the city names. The aggregated column is the average age of the listed inhabitants of this city. In SQL the query would be something like SELECT city,AVG(age) FROM persons GROUP BY city;.

aggr -f name_city_age.csv -d ";" avg:2 1 | sort -rn

Will basically do the same as in the previous example, but outputs the average-age column first and uses sort -rn to output the rows in descending order.

aggr -f name_city_age.csv -d ";" 0 count:*

This time we're using the name column for grouping and count as aggregate function. This gives us the number how often which name is in the list.

Find more examples in the wiki: https://github.com/stefanschramm/aggr/wiki


.SH MAJOR DIFFERENCES TO SQL'S GROUP BY:

In aggr an aggregate function can only be applied to a single column. That means expressions like SUM(price * amount) in SQL are not possible here. To achieve this you should first calculate the product with some other commandline utility into a single column and then pipe it's output to aggr.

.SH AUTHOR
Stefan Schramm <mail@stefanschramm.net>


aggr.py
#!/usr/bin/python -tt
# -*- coding: utf-8 -*-

# aggr.py - aggr - aggregate CSV-data and print on the standard output
#
# Copyright (C) 2010, Stefan Schramm <mail@stefanschramm.net>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


import sys

def parse_cmdline_arguments(args):

    # defaults
    filename = None
    delimiter = "\t"
    outputs = []
    group = []

    # parsing and syntax check
    args = iter(args)
    while True:
        try:
            arg = args.next()
        except StopIteration:
            break
        if arg == '-f':
            # filename
            try:
                filename = args.next()
                continue
            except StopIteration:
                sys.exit("Missing argument for -f (filename)")
        if arg == '-d':
            # delimiter
            try:
                delimiter = args.next()
                continue
            except StopIteration:
                sys.exit("Missing argument for -d (field delimiter)")
        # interpret argument as output column
        outputs.append(arg)
        if arg.isdigit():
            # if it's only a number: group column
            group.append(int(arg))

    return filename, delimiter, outputs, group

def get_data_source(filename):
    if filename != None and filename != "-":
        try:
            lines = open(filename, "r")
        except IOError:
            sys.exit("Unable to open " + filename + " for reading")
    else:
        lines = sys.stdin
    return lines

def collect_groups_with_order(lines, delimiter, group):
    groups = {}
    grouporder = []
    for line in lines:
        fields = line.rstrip().split(delimiter)
        key = []
        for position in group:
            try:
                key.append(fields[position])
            except IndexError:
                sys.exit("Column number " + str(position) +  " out of range")
        key = tuple(key)
        if groups.has_key(key):
            groups[key].append(fields)
        else:
            grouporder.append(key) # remember order of first occurances of groups
            groups[key] = [fields]
    return groups, grouporder

def output_groups(groups, grouporder, outputs, delimiter):
    for group in grouporder:
        output_line = []
        for output in outputs:
            if output.isdigit():
                output_line.append(groups[group][0][int(output)]) # no aggregation
            else:
                output_line.append(aggregate(groups[group], output)) # apply aggregate function
        print delimiter.join(output_line)

def aggregate(data, output):
    
    try:
        function, column = output.split(":")
    except ValueError:
        sys.exit("Syntax error for output column: " + output)
        
    # special case: *,count
    if function == "count" and column == "*":
        return str(len(data))

    # following functions require column to be an integer
    if not column.isdigit():
        sys.exit("Column number for aggregate function must be numeric (or *, but only for count).")
    column = int(column);
        
    try:
        # most aggregate functions follow a simple map-reduce-pattern
        # some don't need map (like first and last) or map in their own way
        if function in ['sum', 'avg', 'max', 'min', 'count']:
            columndata = map(lambda row: row[column], data)

        # make out aggregate function
        if function == 'sum':
            return str(reduce(lambda a, b: a + float(b), columndata, 0))
        if function == 'count':
            # an empty string has the meaning that NULL has in SQL
            return str(len(filter(lambda row: row != "", columndata)))
        if function == 'avg':
            return str(reduce(lambda a, b: a + float(b), columndata, 0) / len(columndata))
        if function == 'max':
            return str(reduce(lambda a, b: max(a,float(b)), columndata, float(columndata[0])))
        if function == 'min':
            return str(reduce(lambda a, b: min(a,float(b)), columndata, float(columndata[0])))
        if function == 'first':
            return data[0][int(column)]
        if function == 'last':
            return data[len(data)-1][int(column)]
        if function == 'median':
            # here we're assuming that a numeric order is wanted!
            columndatasorted = sorted(map(lambda row: float(row[column]), data))
            if len(columndatasorted) % 2 == 0:
                # even: return average of the 2 elements in the middle
                return str((columndatasorted[len(columndatasorted) / 2 - 1] + columndatasorted[len(columndatasorted) / 2]) / 2)
            else:
                # odd: return element in the middle
                return str(columndatasorted[len(columndatasorted) / 2])
        sys.exit("Unknown aggregate function: " + function)
    except IndexError:
        sys.exit("Column number " + str(column) +  " out of range")
    except ValueError:
        sys.exit("Unable to cast input data of column " + str(column) + " to type required by aggregate function " + function)

if __name__ == "__main__":

    try:
        # parse command line options
        (filename, delimiter, outputs, group) = parse_cmdline_arguments(sys.argv[1:])

        # determine data source (file or stdin)
        lines = get_data_source(filename)

        # collect groups with order of first occurance
        groups, grouporder = collect_groups_with_order(lines, delimiter, group)

        # print groups
        output_groups(groups, grouporder, outputs, delimiter)

    except KeyboardInterrupt:
        sys.exit("Aborted by user")
